### PR TITLE
Ángel Vázquez - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by Angel5215**
Xcode 12.2 for the UI Design Patterns Challenge.

### CODE REVIEW

File: MVP Module/Source/Feed Presentation/FeedErrorViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,7 @@
+//
+//  Copyright © 2019 Essential Developer. All rights reserved.
+//
+
+struct FeedErrorViewModel {
+	let errorMessage: String
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedErrorViewModel.swift:6]: The `errorMessage` could be an optional `String?`, so the Presenter can decide when to show/hide the message.

For example, in the `didStartLoadingFeed()` method, the Presenter can call  `feedErrorView.display(.none)` to hide the error.

This way, the ViewController doesn't need to decide when to hide the error. It's often easier to manage state when all changes are centralized.

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 2]
```diff
@@ -8,20 +8,23 @@ protocol FeedViewControllerDelegate {
...
+	@IBOutlet var errorView: ErrorView?
+	
 	public override func viewDidLoad() {
 		super.viewDidLoad()
 		
 		refresh()
 	}
 	
 	@IBAction private func refresh() {
+		errorView?.hideMessage()
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:27]: The Presenter should decide when to show/hide messages. See my other comment.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 3]
```diff
@@ -20,12 +20,16 @@ final class FeedViewModel {
 	
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onFeedLoadError: Observer<String>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: `onFeedLoadError: Observer<String> ` could be `onErrorStateChange: Observer<String?>`, so the ViewModel decides when to show/hide errors (instead of the view controller). It's often easier to manage state when all changes are centralized.

For example, in the `loadFeed()` method, the ViewModel can call  `onErrorStateChange(.none)` to hide the error before loading the feed.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 4]
```diff
@@ -20,12 +20,16 @@ final class FeedViewModel {
...
 	
 	func loadFeed() {
 		onLoadingStateChange?(true)
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:26]: Following my suggestion above, you could update the error state to hide the message before loading again:

`onErrorStateChange(.none)`

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 5]
```diff
@@ -20,6 +22,7 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 	}
 	
 	@IBAction private func refresh() {
+		errorView?.hideMessage()
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:25]: The ViewModel should decide when to show/hide messages. See my other comment.

**Review by caiozullo — COMMENTED** [review]
Hi @Angel5215,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

I left some review comments. Feel free to push changes.

Great job!

**Comment by Angel5215** [comment]
Thanks for the feedback! 

I was unsure how to hide the error message when implementing my solution, but now everything seems more clear to me.

**Comment by caiozullo** [comment]
Nice one!
